### PR TITLE
feat: add upgrade oracle tx & cli

### LINKED ISSUE #571: Implement upgrade oracle tx
- add CLI `upgrade-oracle [unique-id] [node-public-key] [node-public-key-remote-report] [trusted-block-hash] [trusted-block-height]`
- implement oracle registration with the upgraded `uniqueID` when oracle upgrade proposal passed and before reaching the upgrade height
  - oracle in upgrade state can get approval and `oraclePrivKey` from oracles by the `OracleRegistration` procedure
- Implement a function that approves `OracleRegistration` of upgraded oracle

### PR DESCRIPTION
**Pull request by audtlr24**
closes #571.

### CODE REVIEW

File: x/oracle/keeper/upgrade.go [hunk 1]
```diff
@@ -58,3 +58,74 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, info *types.OracleUpgradeInfo) err
 	ctx.Logger().Info("Oracle upgrade was successful.", fmt.Sprintf("uniqueID: %s, height: %v", info.UniqueId, info.Height))
 	return nil
 }
+
+func (k Keeper) UpgradeOracle(ctx sdk.Context, msg *types.MsgUpgradeOracle) error {
+	oracleUpgrade := types.NewUpgradeOracle(msg)
+
+	if err := oracleUpgrade.ValidateBasic(); err != nil {
+		return err
+	}
```
Comment (H4NLee) [inline x/oracle/keeper/upgrade.go:67]: We don't need this validation. This basic validation probably would be done when validate each message (in this case `MsgUpgradeOracle`)
Accordingly, there is no need to have `OracleUpgrade`'s `ValidateBasic()` in `types/oracle.go`.

File: x/oracle/keeper/upgrade.go [hunk 2]
```diff
@@ -58,3 +58,74 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, info *types.OracleUpgradeInfo) err
 	ctx.Logger().Info("Oracle upgrade was successful.", fmt.Sprintf("uniqueID: %s, height: %v", info.UniqueId, info.Height))
 	return nil
 }
+
+func (k Keeper) UpgradeOracle(ctx sdk.Context, msg *types.MsgUpgradeOracle) error {
```
Comment (H4NLee) [inline x/oracle/keeper/upgrade.go:62]: Why don't we do duplication check for upgrade as done in #582?

File: x/oracle/keeper/upgrade.go [hunk 3]
```diff
@@ -58,3 +58,74 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, info *types.OracleUpgradeInfo) err
...
+	)
+	return nil
+}
+
+func (k Keeper) GetOracleUpgrade(ctx sdk.Context, uniqueID, address string) (*types.OracleUpgrade, error) {
+	store := ctx.KVStore(k.storeKey)
+	accAddr, err := sdk.AccAddressFromBech32(address)
+	if err != nil {
+		return nil, err
+	}
```
Comment (H4NLee) [inline x/oracle/keeper/upgrade.go:100]: Please wrap an error with `types.ErrGetOracleUpgrade`.

File: x/oracle/keeper/upgrade.go [hunk 4]
```diff
@@ -58,3 +58,74 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, info *types.OracleUpgradeInfo) err
...
+	key := types.GetOracleUpgradeKey(uniqueID, accAddr)
+	bz := store.Get(key)
+	if bz == nil {
+		return nil, sdkerrors.Wrapf(types.ErrGetOracleUpgrade, "oracle registration not found")
```
Comment (H4NLee) [inline x/oracle/keeper/upgrade.go:104]: We can make a sentinel error of `types.ErrOracleUpgradeNotFound` and return it without wrapping.

File: x/oracle/types/oracle.go [hunk 5]
```diff
@@ -128,3 +128,42 @@ func (m *OracleUpgradeInfo) ShouldExecute(ctx sdk.Context) bool {
...
+		UniqueId:               msg.UniqueId,
+		OracleAddress:          msg.OracleAddress,
+		NodePubKey:             msg.NodePubKey,
+		NodePubKeyRemoteReport: msg.NodePubKeyRemoteReport,
+		TrustedBlockHeight:     msg.TrustedBlockHeight,
+		TrustedBlockHash:       msg.TrustedBlockHash,
+	}
+}
+
+func (m *OracleUpgrade) ValidateBasic() error {
```
Comment (H4NLee) [inline x/oracle/types/oracle.go:143]: As commented, we probably don't need this validation

**Review by H4NLee — COMMENTED** [review]
Thx. I left some comments.

File: x/oracle/types/oracle.go (hunk 5 above)
Comment (H4NLee) [reply x/oracle/types/oracle.go:143]: You can ignore this comment. This validation can be used for genesis state.

File: x/oracle/keeper/upgrade.go (hunk 2 above)
Comment (audtlr24) [reply x/oracle/keeper/upgrade.go:62]: Can you tell me what kind of duplication check it is? Thanks

Comment (H4NLee) [reply x/oracle/keeper/upgrade.go:62]: I meant that let's check if the oracle has already requested upgrade.

The exact link for example is:
https://github.com/medibloc/panacea-core/blob/1c69a6de0ebc97888b33d92718875dc523c84cac/x/oracle/keeper/oracle.go#L20

Maybe,
```go
existing, err := k.GetOracleUpgrade(ctx, msg.GetUniqueId(), msg.GetOracleAddress())
// check already requested upgrade
if existing {
    return fmt.Errorf("already requested upgrade")
}
```

Comment (audtlr24) [reply x/oracle/keeper/upgrade.go:63]: Okay, thanks. I fixed it.

**Review by inchori — APPROVED** [review]
LGTM!

File: x/oracle/keeper/upgrade_test.go [hunk 6]
```diff
@@ -52,3 +76,117 @@ func (suite *oracleUpgradeTestSuite) TestApplyUpgradeSuccess() {
...
+		TrustedBlockHeight:     int64(1),
+		TrustedBlockHash:       []byte("trustedBlockHash"),
+	}
+
+	err := suite.OracleKeeper.UpgradeOracle(ctx, msgOracleUpgrade)
+	suite.Require().Error(err, types.ErrUpgradeOracle)
+	suite.Require().ErrorContains(err, "does not match the upgrade uniqueID")
+}
+
+func (suite *oracleUpgradeTestSuite) TestUpgradeOracleFaileNoRegisteredOracle() {
```
Comment (H4NLee) [inline x/oracle/keeper/upgrade_test.go:170]:
```suggestion
func (suite *oracleUpgradeTestSuite) TestUpgradeOracleFailedNoRegisteredOracle() {
```

**Review by H4NLee — APPROVED** [review]
lgtm

**Review by gyuguen — APPROVED** [review]
LGTM

**Review by youngjoon-lee — COMMENTED** [review]
lgtm

**Pull request merged**